### PR TITLE
feat: support using existing identity for addons

### LINKED ISSUE #2233: Support associating existing user assigned identity to nodes
**Describe the request**
Support associating existing user assigned identity to nodes. Currently AKS-Engine only supports creating a new one when creating the cluster.
**Explain why AKS Engine needs it**
We're working on switching all addons in AKS to use user assigned identity to authenticate, and AKS RP will create one user assigned identity for each addon. After this, we need to associate the identity to every agent node in the cluster, it would be very convenient if AKS-Engine supports associating existing identity to agent nodes.
**Describe the solution you'd like**
A new struct containing the resource id of existing user assigned identities will be added to “ContainerService” and specified identities will be associated to every (agent) node in generated template. 
**Describe alternatives you've considered**
Associating identities can also be done by AKS RP, but this way has a big shortcoming: when upgrading the cluster, for a newly created VM, identities will not be associated with it. So if addon pod is scheduled to a newly created VM, it can't work correctly for a period of time.
**Additional context**

### PR DESCRIPTION
**Pull request by norshtein**
**Reason for Change**:

Detailed background can be found in #2233 

**Issue Fixed**:

Fixes #2233 

**Requirements**:

- [X] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:
 - Though AKS will only generate agent nodes' ARM template, I associate addons' identity to master VM(SS) to keep consistent with agent nodes' behaviror.

### CODE REVIEW

**Comment by norshtein** [comment]
I noticed when managed identity is enabled, one extension will be installed on master VMSS(there is no such extension in agent VMSS): https://github.com/Azure/aks-engine/blob/master/pkg/engine/virtualmachinescalesets.go#L274-L289 . What's the purpose of the extension? @jackfrancis

File: pkg/api/types.go [hunk 1]
```diff
@@ -109,6 +109,16 @@ type ClusterMetadata struct {
 type AddonProfile struct {
 	Enabled bool              `json:"enabled"`
 	Config  map[string]string `json:"config"`
+	// Identity contains information of the identity associated with this addon.
+	// This property will only appear in an MSI-enabled cluster.
+	Identity *UserAssignedIdentity `json:"identity,omitempty"`
```
Comment (kkmsft) [inline pkg/api/types.go]: The add-on facility used in aks-engine is different from the add-on facility used in AKS. Hence I am not sure how this can be used in AKS. I was expecting the facility to be more generic and be VMSS/VMA template specific. What is the missing piece piece connecting this ?

File: pkg/engine/virtualmachinescalesets.go [hunk 2]
```diff
@@ -440,6 +457,21 @@ func CreateAgentVMSS(cs *api.ContainerService, profile *api.AgentPoolProfile) Vi
 		}
 	}
 
+	for _, addonProfile := range addonProfiles {
+		if addonProfile.Enabled && addonProfile.Identity != nil && addonProfile.Identity.ResourceID != "" {
+			if virtualMachineScaleSet.Identity == nil {
+				virtualMachineScaleSet.Identity = &compute.VirtualMachineScaleSetIdentity{}
+				virtualMachineScaleSet.Identity.Type = compute.ResourceIdentityTypeUserAssigned
+				virtualMachineScaleSet.Identity.UserAssignedIdentities = make(map[string]*compute.VirtualMachineScaleSetIdentityUserAssignedIdentitiesValue)
```
Comment (kkmsft) [inline pkg/engine/virtualmachinescalesets.go]: Can we make it a single step by using an initialization which is already filled with Type and UserAssignedIdentities ?

File: pkg/engine/virtualmachines.go [hunk 3]
```diff
@@ -405,6 +422,21 @@ func createAgentAvailabilitySetVM(cs *api.ContainerService, profile *api.AgentPo
 		}
 	}
 
+	for _, addonProfile := range addonProfiles {
+		if addonProfile.Enabled && addonProfile.Identity != nil && addonProfile.Identity.ResourceID != "" {
+			if virtualMachine.Identity == nil {
+				virtualMachine.Identity = &compute.VirtualMachineIdentity{}
+				virtualMachine.Identity.Type = compute.ResourceIdentityTypeUserAssigned
+				virtualMachine.Identity.UserAssignedIdentities = make(map[string]*compute.VirtualMachineIdentityUserAssignedIdentitiesValue)
+			} else if virtualMachine.Identity.Type == compute.ResourceIdentityTypeSystemAssigned {
```
Comment (kkmsft) [inline pkg/engine/virtualmachines.go]: Aren't there other possibilities such as:
1. Identity.Type is None ?
2. Identity.Type is already ResourceIdentityTypeUserAssigned? 

Can we please handle/add comments on how those scenarios works ?

File: pkg/engine/virtualmachinescalesets.go [hunk 4]
```diff
@@ -98,6 +99,21 @@ func CreateMasterVMSS(cs *api.ContainerService) VirtualMachineScaleSetARM {
 		virtualMachine.Identity = identity
 	}
 
+	for _, addonProfile := range addonProfiles {
+		if addonProfile.Enabled && addonProfile.Identity != nil && addonProfile.Identity.ResourceID != "" {
+			if virtualMachine.Identity == nil {
+				virtualMachine.Identity = &compute.VirtualMachineScaleSetIdentity{}
```
Comment (kkmsft) [inline pkg/engine/virtualmachinescalesets.go]: Is there a possibility to move the Identity handling to a utility function and reuse it in VMA and VMSS case ?

**Comment by ritazh** [comment]
> one extension will be installed on master VMSS(there is no such extension in agent VMSS)

The `ManagedIdentityExtensionForLinux` extension was deprecated in Jan 2019 and was removed in this PR: https://github.com/Azure/aks-engine/pull/185. But it was added back in a subsequent PR: https://github.com/Azure/aks-engine/blame/master/pkg/engine/virtualmachinescalesets.go#L274-L289

File: pkg/api/types.go (hunk 1 above)
Comment (norshtein) [reply pkg/api/types.go]: When generating ARM template, AKS RP will also pass `aksEngineAPI.AddonProfile` to AKS-Engine. After adding this in AKS-Engine, AKS RP will add the resourceID of created identity in `AddonProfile.Identity` so that AKS-Engine can associate the identity to every agent node.

File: pkg/engine/virtualmachines.go (hunk 3 above)
Comment (norshtein) [reply pkg/engine/virtualmachines.go]: Thanks for reminding, will add comment.

File: pkg/api/types.go (hunk 1 above)
Comment (kkmsft) [reply pkg/api/types.go]: Was wondering won't the deployment of the template add the add-on enabled to add-on manager and then add-on manager pull the corresponding images when you deploy the template ? Wouldn't that interfere with the helm installation of add-on in case of AKS ?

File: pkg/engine/virtualmachinescalesets.go (hunk 2 above)
Comment (norshtein) [reply pkg/engine/virtualmachinescalesets.go]: Thanks for reminding, updated.

File: pkg/engine/virtualmachinescalesets.go (hunk 4 above)
Comment (norshtein) [reply pkg/engine/virtualmachinescalesets.go]: Thanks for reminding, it makes sense, updated.

**Comment by norshtein** [comment]
> > one extension will be installed on master VMSS(there is no such extension in agent VMSS)
> 
> The `ManagedIdentityExtensionForLinux` extension was deprecated in Jan 2019 and was removed in this PR: #185. But it was added back in a subsequent PR: https://github.com/Azure/aks-engine/blame/master/pkg/engine/virtualmachinescalesets.go#L274-L289
> 
> This should probably be removed.

Thanks for the info. Just removed the extension.

File: pkg/api/types.go (hunk 1 above)
Comment (norshtein) [reply pkg/api/types.go]: > wouldn't the deployment of the template add the enabled aks-engine add-ons to add-on manager and then pull the corresponding images when deployed?

I guess not? I haven't seen such code in AKS-Engine. The deployment of the template will do nothing related to addon. And seems in current AKS-Engine, it just ignores all `AddonProfile`. @jackfrancis @ritazh  could you please help confirm? If I understand this correctly, the management of all addons is fully controlled by AKS RP.

Comment (norshtein) [reply pkg/api/types.go]: Seems AKS's addon is recorded in `ContainerService.Properties.AddonProfiles`, AKS-Engine's addon is recorded in `ContainerService.Properties.OrchestratorProfile.KubernetesConfig.Addons`. Currently AKS-Engine will do nothing about addons in `ContainerService.Properties.AddonProfiles`.

**Comment by norshtein** [comment]
Hey @jackfrancis , @yangl900 , could you please have a review?

File: pkg/engine/virtualmachinescalesets.go [hunk 5]
```diff
@@ -271,23 +273,6 @@ func CreateMasterVMSS(cs *api.ContainerService) VirtualMachineScaleSetARM {
 
 	var extensions []compute.VirtualMachineScaleSetExtension
 
-	if useManagedIdentity {
```
Comment (jackfrancis) [inline pkg/engine/virtualmachinescalesets.go]: Why are we getting rid of this? This breaks the following cluster configuration scenario:

```
cs.Properties.OrchestratorProfile.KubernetesConfig.UseManagedIdentity = true
cs.Properties.OrchestratorProfile.KubernetesConfig.UserAssignedID = "fooAssignedID"
```

Am I missing something?

@ritazh @sozercan Am I correct in recalling that one of you introduced MSI into aks-engine?

Comment (sozercan) [reply pkg/engine/virtualmachinescalesets.go]: Wasn't me, but I believe the extension is deprecated in favor of IMDS.
https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/howto-migrate-vm-extension

Comment (jackfrancis) [reply pkg/engine/virtualmachinescalesets.go]: Thanks @sozercan, that suggests this deletion actually fixes a bug? Should we do that as a separate PR so we can track any regressions distinctly?

Comment (sozercan) [reply pkg/engine/virtualmachinescalesets.go]: missed this, looks like @ritazh commented earlier: https://github.com/Azure/aks-engine/pull/2238#issuecomment-548163703

Comment (jackfrancis) [reply pkg/engine/virtualmachinescalesets.go]: Ah, thank you @ritazh and @sozercan, I should read the comment history in addition to the changeset. :)

I think ideally this would have been fixed in a distinct PR, but we don't have to be strict here. Looks like a good change that helped to clarify the scope of work here.

File: pkg/engine/virtualmachines.go [hunk 6]
```diff
@@ -601,3 +605,33 @@ func addCustomTagsToVM(tags map[string]string, vm *compute.VirtualMachine) {
 		}
 	}
 }
+
+func associateAddonIdentitiesToVM(addonProfiles map[string]api.AddonProfile, virtualMachine *compute.VirtualMachine) {
```
Comment (jackfrancis) [inline pkg/engine/virtualmachines.go]: We need UT for this new func. Please add a test case for a nil `addonProfiles` and `virtualMachine`. Which means that we'll need to exit the func right away if a nil `virtualMachine` input is passed in.

File: pkg/engine/virtualmachinescalesets.go [hunk 7]
```diff
@@ -828,3 +815,33 @@ func addCustomTagsToVMScaleSets(tags map[string]string, vm *compute.VirtualMachi
 		}
 	}
 }
+
+func associateAddonIdentitiesToVMSS(addonProfiles map[string]api.AddonProfile, virtualMachineScaleSet *compute.VirtualMachineScaleSet) {
```
Comment (jackfrancis) [inline pkg/engine/virtualmachinescalesets.go]: ditto above comment about UT

File: pkg/api/types.go (hunk 1 above)
Comment (ritazh) [reply pkg/api/types.go]: I have a similar question as @kkmsft. The changes you are making updates the aks-engine addons, which will be deployed by the add-on manager. How does this work with AKS addons? Can you pls verify this?

Comment (jackfrancis) [reply pkg/api/types.go]: The `AddonProfile` struct is actually not used by aks-engine. aks-engine uses the `KubernetesAddon` struct which is composed as a child property of `KubernetesConfig`.

File: pkg/engine/virtualmachines.go (hunk 6 above)
Comment (norshtein) [reply pkg/engine/virtualmachines.go]: Thanks for reviewing! added unit test.

File: pkg/engine/virtualmachinescalesets.go (hunk 7 above)
Comment (norshtein) [reply pkg/engine/virtualmachinescalesets.go]: added :D

**Comment by norshtein** [comment]
Hi @jackfrancis , is there any ETA of a new release containing this PR?

**Comment by norshtein** [comment]
Rebased upon latest master.

File: pkg/engine/virtualmachinescalesets.go [hunk 8]
```diff
@@ -828,3 +815,36 @@ func addCustomTagsToVMScaleSets(tags map[string]string, vm *compute.VirtualMachi
...
+			if virtualMachineScaleSet.Identity == nil {
+				virtualMachineScaleSet.Identity = &compute.VirtualMachineScaleSetIdentity{
+					Type:                   compute.ResourceIdentityTypeUserAssigned,
+					UserAssignedIdentities: make(map[string]*compute.VirtualMachineScaleSetIdentityUserAssignedIdentitiesValue),
+				}
+			} else if virtualMachineScaleSet.Identity.Type == compute.ResourceIdentityTypeSystemAssigned {
+				virtualMachineScaleSet.Identity.Type = compute.ResourceIdentityTypeSystemAssignedUserAssigned
+				virtualMachineScaleSet.Identity.UserAssignedIdentities = make(map[string]*compute.VirtualMachineScaleSetIdentityUserAssignedIdentitiesValue)
+			} else if virtualMachineScaleSet.Identity.Type == compute.ResourceIdentityTypeNone {
+				// Note: in current AKS-Engine's implementation, we will never enter into this branch. Just handle it here in case implementation
```
Comment (xuto2) [inline pkg/engine/virtualmachinescalesets.go]: late comment: why not return error in this case?

Comment (xuto2) [reply pkg/engine/virtualmachinescalesets.go]: at least we should log some message?

Comment (norshtein) [reply pkg/engine/virtualmachinescalesets.go]: Yeah, logging a warning here is better, but there is no logger available in current context. As for return an error, as the function name `associateAddonIdentitiesToVMSS` indicates, this function should associate identities to VMSS, if `virtualMachineScaleSet.Identity.Type` is `None`(which will never happen in current AKS-Engine's implementation but may happen later), instead of returning error, just change `virtualMachineScaleSet.Identity.Type` to `UserAssigned` and associate addons' identities to the VMSS makes more sense.
